File: CSES/Additional Problems/P1740.java
//make sure to make new file!
import java.io.*;
import java.util.*;

public class P1740{

   public static int N = 2000001;
   public static long[] bits;
   public static int O = 1000001;
   
   public static void main(String[] args)throws IOException{
      FastScanner fs = new FastScanner();
      PrintWriter out = new PrintWriter(System.out);
      
      int n = fs.nextInt();
      
      ArrayList<ArrayList<Integer>> inc = new ArrayList<ArrayList<Integer>>(N+1);
      ArrayList<ArrayList<Integer>> dec = new ArrayList<ArrayList<Integer>>(N+1);
      for(int k = 0; k <= N; k++){
         inc.add(new ArrayList<Integer>());
         dec.add(new ArrayList<Integer>());
      }
      
      ArrayList<Vert> verts = new ArrayList<Vert>();
      
      for(int k = 0; k < n; k++){
         int x1 = fs.nextInt()+O;
         int y1 = fs.nextInt()+O;
         int x2 = fs.nextInt()+O;
         int y2 = fs.nextInt()+O;
         
         if(x1 == x2 && y1 == y2) continue;
         if(x1 == x2){
            verts.add(new Vert(x1,y1,y2));
         } else if(y1 == y2){
            int xmin = Math.min(x1,x2);
            int xmax = Math.max(x1,x2);
            
            inc.get(xmin).add(y1);
            dec.get(xmax).add(y1);
         }  
      }
      
      Collections.sort(verts);
      
      bits = new long[N];
      
      long answer = 0L;
      int vi = 0;
      for(int x = 1; x <= N; x++){
         while(vi < verts.size() && verts.get(vi).x == x){
            answer += psum(verts.get(vi).y2) - psum(verts.get(vi).y1-1);
            vi++;
         }
         
         for(int i : inc.get(x)){
            update(i,1L);
         }
         for(int i : dec.get(x)){
            update(i,-1L);
         }
      }
      
      out.println(answer);
      
      
      
      
      
      
      
      
      
      out.close();
   }
   
   public static class Vert implements Comparable<Vert>{
      int x;
      int y1;
      int y2;
      
      public Vert(int a, int b, int c){
         x = a;
         y1 = Math.min(b,c);
         y2 = Math.max(b,c);
      }
      
      public int compareTo(Vert v){
         return x-v.x;
      }
   }
      
   public static void update(int i, long x){
      for(; i <= N; i += i&-i){
         //System.out.println(i);
         bits[i]+=x;
      }
   
   }
   
   public static long psum(int i){
      long sum = 0L;
      for(; i > 0; i -= i&-i){
         //System.out.println(i);
         sum += bits[i];
      }
      return sum;
   
   }

      
      
}


class FastScanner
{
    //I don't understand how this works lmao
    private int BS = 1 << 16;
    private char NC = (char) 0;
    private byte[] buf = new byte[BS];
    private int bId = 0, size = 0;
    private char c = NC;
    private double cnt = 1;
    private BufferedInputStream in;
 
    public FastScanner() {
        in = new BufferedInputStream(System.in, BS);
    }
 
    public FastScanner(String s) {
        try {
            in = new BufferedInputStream(new FileInputStream(new File(s)), BS);
        } catch (Exception e) {
            in = new BufferedInputStream(System.in, BS);
        }
    }
 
    private char getChar() {
        while (bId == size) {
            try {
                size = in.read(buf);
            } catch (Exception e) {
                return NC;
            }
            if (size == -1) return NC;
            bId = 0;
        }
        return (char) buf[bId++];
    }
 
    public int nextInt() {
        return (int) nextLong();
    }
 
    public int[] nextInts(int N) {
        int[] res = new int[N];
        for (int i = 0; i < N; i++) {
            res[i] = (int) nextLong();
        }
        return res;
    }
 
    public long[] nextLongs(int N) {
        long[] res = new long[N+1];
        for (int i = 1; i <= N; i++) {
            res[i] = nextLong();
        }
        return res;
    }
 
    public long nextLong() {
        cnt = 1;
        boolean neg = false;
        if (c == NC) c = getChar();
        for (; (c < '0' || c > '9'); c = getChar()) {
            if (c == '-') neg = true;
        }
        long res = 0;
        for (; c >= '0' && c <= '9'; c = getChar()) {
            res = (res << 3) + (res << 1) + c - '0';
            cnt *= 10;
        }
        return neg ? -res : res;
    }
 
    public double nextDouble() {
        double cur = nextLong();
        return c != '.' ? cur : cur + nextLong() / cnt;
    }
 
    public double[] nextDoubles(int N) {
        double[] res = new double[N];
        for (int i = 0; i < N; i++) {
            res[i] = nextDouble();
        }
        return res;
    }
 
    public String next() {
        StringBuilder res = new StringBuilder();
        while (c <= 32) c = getChar();
        while (c > 32) {
            res.append(c);
            c = getChar();
        }
        return res.toString();
    }
 
    public String nextLine() {
        StringBuilder res = new StringBuilder();
        while (c <= 32) c = getChar();
        while (c != '\n') {
            res.append(c);
            c = getChar();
        }
        return res.toString();
    }
 
    public boolean hasNext() {
        if (c > 32) return true;
        while (true) {
            c = getChar();
            if (c == NC) return false;
            else if (c > 32) return true;
        }
    }
}

File: CSES/Additional Problems/P1740.cpp
#include <bits/stdc++.h>

using namespace std;

typedef pair<int,pair<int,int>> Vert;

Vert make_vert(int x, int y1, int y2){
   return make_pair(x,make_pair(min(y1,y2),max(y1,y2)));
}

int N = 2000001;
int O = 1000001;
vector<long long> bits;

void update(int i, long long x){
   for(; i <= N; i += i&-i){
      bits[i]+=x;
   }
}

long long psum(int i){
   long long cursum = 0LL;
   for(; i > 0; i -= i&-i){
      cursum += bits[i];
   }
   return cursum;
}


int main(){
   ios::sync_with_stdio(false);
   cin.tie(0);
   
   int n;
   cin >> n;
   
   vector<vector<int>> inc(N+1,vector<int>());
   vector<vector<int>> dec(N+1,vector<int>());
   
   vector<Vert> verts;
   
   for(int k = 0; k < n; k++){
      int x1,y1,x2,y2;
      cin >> x1 >> y1 >> x2 >> y2;
      x1 += O;
      y1 += O;
      x2 += O;
      y2 += O;
      
      if(x1 == x2 && y1 == y2) continue;
      if(x1 == x2){
         verts.push_back(make_vert(x1,y1,y2));
      } else if(y1 == y2){
         int xmin = min(x1,x2);
         int xmax = max(x1,x2);
         
         inc[xmin].push_back(y1);
         dec[xmax].push_back(y1);
      }
   }
   
   sort(verts.begin(),verts.end());
   
   bits = vector<long long>(N+1,0LL);
   
   long long answer = 0LL;
   int vi = 0;
   for(int x = 1; x <= N; x++){
      while(vi < verts.size() && verts[vi].first == x){
         answer += psum(verts[vi].second.second) - psum(verts[vi].second.first-1);
         vi++;
      }
      
      for(int i : inc[x]){
         update(i,1LL);
      }
      for(int i : dec[x]){
         update(i,-1LL);
      }
   }
   
   cout << answer << endl;
         
   
   
   return 0;
}